File: Dot examples/ASP Book Examples/Book Examples/Chapter34/cs/App_Code/Entities/EntryTag.cs
using System;
using System.Linq;
using System.Data.Linq.Mapping;
using System.Collections;
using System.Collections.Generic;
using System.Web;

/// <summary>
/// Summary description for Tag
/// </summary>
public partial class EntryTag: EntityBase<EntryTag>
{

    public IEnumerable<EntryTag> SelectByEntryId(int entryId)
    {
        return Table.Where( t => t.EntryId == entryId );
    }


    public static IEnumerable<CloudItem> SelectTagCloud()
    {
        return Table.GroupBy( t => t.Name ).Select( g => new CloudItem{Name = g.Key, Count= g.Count()} ).OrderBy( a => a.Name );
    }



    public override void Insert()
    {
        int count = Table.Where( t => t.EntryId == EntryId ).Count();
        if (count < 3)
            base.Insert();
        //else 
        //    ValidationErrors.AddAndThrow("Name", "You cannot add more than 3 tags");
    }


    public static string[] GetSuggestions(string prefixText, int count)
    {
        return Table.Where(t => t.Name.StartsWith(prefixText)).Select(t => t.Name).Distinct().Take(count).ToArray();
    }




    protected override void Validate()
    {
        // Security check
        if (!ValidationUtility.IsInRole("Administrators"))
        {
            string userName = HttpContext.Current.User.Identity.Name;
            Entry entry = Entry.Table.Where( e => e.Id == EntryId && e.AuthorUserName == userName ).SingleOrDefault();
            if (entry == null)
                ValidationErrors.Add("AuthorUserName", "Only the original author or an administrator can modify these tags");
        }

 
        if (!ValidationUtility.SatisfiesRequired(Name))
            ValidationErrors.Add("Name", "Required");    
    }
}

File: Dot examples/ASP Book Examples/Book Examples/Chapter34/vb/App_Code/Entities/EntryTag.vb
Imports System
Imports System.Linq
Imports System.Data.Linq.Mapping
Imports System.Collections
Imports System.Collections.Generic
Imports System.Web

''' <summary>
''' Summary description for Tag
''' </summary>
Partial Public Class EntryTag
     Inherits EntityBase(Of EntryTag)

    Public Function SelectByEntryId(ByVal enTryId As Integer) As IEnumerable(Of EntryTag)
        Return Table _
            .Where(Function(t) t.EntryId = enTryId)
    End Function


    Public Shared Function SelectTagCloud() As IEnumerable(Of CloudItem)
        Return Table _
            .GroupBy(Function(t) t.Name) _
            .Select(Function(g) New CloudItem() With {.Name = g.Key, .Count = g.Count()}) _
            .OrderBy(Function(a) a.Name)
    End Function



    Public Overrides Sub Insert()
        Dim count As Integer = Table _
            .Where(Function(t) t.EntryId = EntryId) _
            .Count()
        If count < 3 Then
            MyBase.Insert()
        Else
            ValidationErrors.Add("Name", "You cannot add more than 3 tags")
        End If
    End Sub


    Public Shared Function GetSuggestions(ByVal prefixText As String, ByVal count As Integer) As String()
        Return Table _
            .Where(Function(t) t.Name.StartsWith(prefixText)) _
            .Select(Function(t) t.Name) _
            .Distinct() _
            .Take(count) _
            .ToArray()
    End Function




    Protected Overrides Sub Validate()
        ' Security check
        If Not ValidationUtility.IsInRole("Administrators") Then
            Dim userName As String = HttpContext.Current.User.Identity.Name
            Dim entry As Entry = entry.Table _
                .Where(Function(e) e.Id = EntryId And e.AuthorUserName = userName) _
                .SingleOrDefault()
            If entry Is Nothing Then
                ValidationErrors.Add("AuthorUserName", "Only the original author or an administrator can modify these tags")
            End If
        End If


        If Not ValidationUtility.SatisfiesRequired(Name) Then
            ValidationErrors.Add("Name", "Required")
        End If
    End Sub
End Class

